### PR TITLE
Market orders

### LINKED ISSUE #105: Market Orders

### PR DESCRIPTION
**Pull request by michael1011**
Closes #105 
I also added market orders to the gRPC interface and the cli.

Although I installed the correct version (3.4.0) of `protoc` there are many unnecessary changes to protobuffs. The only difference was that I compiled it on my laptop this time. @sangaman

### CODE REVIEW

File: lib/service/Service.ts [hunk 1]
```diff
@@ -141,14 +141,22 @@ class Service {
   }
 
   /**
-   * Add an order to the order book.
+   * Add a limit order to the order book.
    */
-  public placeOrder = async (order: OwnOrder) => {
```
Comment (moshababo) [inline lib/service/Service.ts]: I think we could keep the 2 calls (limit/market) unified on `gRPC`/`Service`, and let them split only when `OrderBook` API is being called. same as you did on CLI, based on if price is specified or not.

@sangaman @kilrau

File: lib/orderbook/OrderBook.ts [hunk 2]
```diff
@@ -80,18 +80,24 @@ class OrderBook {
     return this.addOwnOrder(order);
   }
 
-  public addMarketOrder = async (_order: orders.OwnOrder) => {
-    // TODO: implement
+  public addMarketOrder = async (_order: orders.MarketOrder): Promise<matchingEngine.OrderMatch[]> => {
```
Comment (moshababo) [inline lib/orderbook/OrderBook.ts]: the `_` in `_order` can be removed. It was used only when the argument wasn't in use, to relax the compiler.

File: lib/orderbook/OrderBook.ts [hunk 3]
```diff
@@ -80,18 +80,24 @@ class OrderBook {
...
+    const price = _order.quantity > 0 ? Number.MAX_VALUE : 0;
+    const order: orders.OwnOrder = { ..._order, price };
+
+    const result = await this.addOwnOrder(order, true);
+    delete result.remainingOrder;
```
Comment (moshababo) [inline lib/orderbook/OrderBook.ts]: The remaining order is being discarded (not added to queues, not broadcasted), but I think we should return it to the client anyway.

**Comment by sangaman** [comment]
@michael1011 What differences are there between your laptop and the other machine you used to run the proto script? Diifferent OS? Just wondering so we can try to pin down what's causing the files to be generated differently.

**Comment by michael1011** [comment]
@sangaman I have no idea. Everything is exactly the same as far as I know. Both times I installed everything as you described it in the [README](https://github.com/ExchangeUnion/xud/blob/master/proto/README.md).

**Comment by michael1011** [comment]
I unified the two calls into one but the cli doesn't work because of inconsistency in the database:

`Error: 13 INTERNAL: Illegal value for createdAtvalue element of type int64: object (not convertible to Long)`

If a order is loaded from the database the data is stored like this `2018-07-05 18:05:43` and not in the unix time format. And the `hostId` in the database is `null` but gRPC does not support a `int32` to be `null`.

Should I take of this issues in this PR?

**Comment by moshababo** [comment]
I think we shouldn't spend time on fixing this, since we want to stop storing orders in database anyway. The `hostId` issue seems to throw an exception, so I would suggest to implement the removal of the orders persistence first: https://github.com/ExchangeUnion/xud/issues/141
It's on @sangaman now, but I don't think he started working on it. So you can grab it instead, and we'll merge this PR after it will be done.

**Comment by michael1011** [comment]
I messed up the branch really bad by rebasing to master and not `origin/master` and force pushing changes which resulted in getting this branch deleted. I will open another PR about market orders ASAP. Sorry guys.

**Comment by moshababo** [comment]
No worries.

**Comment by sangaman** [comment]
I've done that myself, it's a good lesson in git.

**Pull request closed without merging**